### ISSUE TITLE
Option to add layer via JSON copy & paste

### ISSUE THREAD

**Issue by muesliq** [maputnik/editor#109]
Many use cases include copying a layer from one stylesheet to another. The current dialogue doesn't make that easy:

![screen shot 2017-03-21 at 11 57 05](https://cloud.githubusercontent.com/assets/1606618/24145474/8adb79b6-0e31-11e7-84ca-d17986ff0bb6.png)

It would be great to have an option there to add a whole JSON chunk.

**Comment by orangemug** [comment]
From my perspective, it would be great to find out a little more about the use cases for this, for a bit of context and to make sure we're solving the right problem. 

My initial thoughts are that maybe we should just remove the modal entirely and just land the user in the 'Layer panel' with an 'Untitled Layer' because from there they can edit any of those properties and also have access to the 'JSON editor' by default, thoughts?

**Comment by muesliq** [comment]
> From my perspective, it would be great to find out a little more about the use cases for this, for a bit of context and to make sure we're solving the right problem.

Use case: You have a layer in Stylesheet A and want to copy it to Stylesheet B. You can easily do that in any text editor but not in Maputnik. (Unless a JSON tab exists as you have proposed somewhere else I think.)

> My initial thoughts are that maybe we should just remove the modal entirely and just land the user in the 'Layer panel' with an 'Untitled Layer' because from there they can edit any of those properties and also have access to the 'JSON editor' by default, thoughts?

Layers look very different depending on their type, that's what makes the modal useful. How would the layout of an "Untitled layer" look?

**Comment by orangemug** [comment]
The _type_ would default to `none` with all other fields hidden (except JSON editor). Then you can modify the _type_ field to get started with your new layer. At this point it would also be valid to just paste a JSON blob into the 'JSON editor'.

Note that

 - The blank layer doesn't appear in the layer list until the type is set
 - The id field would be hidden until the type is set, and defaults to `untitled_{NUM}`
 - The layer isn't saved until the layer type is set, navigating away will leave no trace of the layer.

I think this solution

 - Reduces the UI, one less modal
 - Same UI to add / edit layer

Thoughts?

**Comment by lukasmartinelli** [comment]
Sorry to always be the boring voice here. But that's again a feature which is too specific imho.

If we have a plain JSON representation/editor of the stylesheet directly in the editor that would solve this use case as well.

I for my part solved this issue by just editing a local style JSON (the Maputnik CLI and propagate the change back).

**Comment by orangemug** [comment]
Hey @lukasmartinelli, I feel like removing the model actually makes things simpler all round. You basically jump straight into the edit view, so get all the features you'd expect right away (including the JSON editor), rather than hiding them behind a modal.

**Comment by muesliq** [comment]
Wouldn't `"type":"none"` break the GL spec, @orangemug? So far everything Maputnik does conforms to the spec.

**Comment by orangemug** [comment]
@muesliq apologies that was unclear. `none` is just a place holder really, until the type is set to something else other than `none` it won't appear in any saved output. 

Anyway I think you make a good point, looking at the spec `id` is the only thing required. So it becomes even easier, we create a layer called `untitled_{NUM}` with a unset type when we press 'Add Layer'. Then we display the edit layer pane. Nothing else _should_ need to change as we can modify the layer type from that pane, or alternatively input JSON.

Note: This also helps with #97 because we can now add a known bottom margin to the layers panel giving us guaranteed space (lets say 200px) to display autocomplete menus.

**Comment by muesliq** [comment]
Ok, that sounds great then. All up for it.

**Comment by lukasmartinelli** [comment]
> Hey @lukasmartinelli, I feel like removing the model actually makes things simpler all round. You basically jump straight into the edit view, so get all the features you'd expect right away (including the JSON editor), rather than hiding them behind a modal.

That's true. I never liked the modal. Go for it.

**Comment by JiaXingKing** [comment]
Hey @lukasmartinelli,I'm new as a map designer. when i used this tool, i was always confused about the source-layer option. at the beginning, I drag the Geojosn or any other source in. but there is no any layer name in sources, just has FeatureCollection. so how can I put a layer name in source-layer option?

**Comment by gregorywolanski** [comment]
I’m working on getting rid of the modal.

**Comment by gregorywolanski** [comment]
Do you prefer to have `LayerEditor` modified or do you want me to establish `LayerCreator` component? @orangemug?

**Comment by orangemug** [comment]
Hey @gregorywolanski sorry for the delay I forgot about this ticket. Whatever is easier really, I'm personally still not 100% how this is going to feel UX wise. However I suggest getting a branch/pull request in a basic working state so we can test it out.

**Comment by gregorywolanski** [comment]
I give up on this issue. 

The changes required in the codebase are too big for me to make as a casual contributor.

I’ll continue to watch the repo and possibly solve some smaller future issues.

**Comment by orangemug** [comment]
Yeah this is quite a big feature @gregorywolanski thanks for taking a look. There will be a bunch of smaller issues raised soon(ish)
